### ISSUE TITLE
[HOLD for payment 2023-08-08] [$1000] Web - An appropriate message is not shown when inviting a member that is already invited.

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#21193]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Go to Settings
2. Go to workspace and choose an existing workspace with invited members
3. Click the invite button and try inviting a member that is already invited
4. Notice that no message appear indicating that the member is already invited.

## Expected Result:
Show`[email address] is already a member of [Workspace name]'

~Appropriate message should appear indicating the member is already invited~

## Actual Result:
No message appeared indicating that the member is already invited.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.27-6

**Reproducible in staging?:** y

**Reproducible in production?:** y

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/93399543/0be1a1a3-a165-4dc0-896f-72484c0323d0

**Expensify/Expensify Issue URL:**

**Issue reported by:** @daveSeife

**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1686728772322259

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0129bb1fb0fcd4cb51</li>
        <li>Upwork Job ID: 1671667770831769600</li>
        2023-07-12<li>Automatic offers: </li>
<ul>
<li>fedirjh | Reviewer | 25717350</li>
<li>eh2077 | Contributor | 25717351</li>
<li>daveSeife | Reporter | 25717352</li>
</ul></li>
    </ul>
</details>

**Comment by mallenexpensify** [comment]
Was able to reproduce, nice find @daveSeife
Pretty sure it can be External
Instead of `No results found`, let's show
`[email address] is already a member of [Workspace name]'
I'll update the OP with this proposed copy too

**Comment by sophiepintoraetz** [comment]
@eh2077 made a [proposal](https://github.com/Expensify/App/issues/21531#issuecomment-1606156784) on this dupe as well.

**Comment by mallenexpensify** [comment]
@JmillsExpensify when you get around to seeing this, can you reply to the above comment? 
- https://github.com/Expensify/App/issues/21193#issuecomment-1607616798

**Comment by mallenexpensify** [comment]
Pinged @JmillsExpensify in NewDot to take a look

**Comment by JmillsExpensify** [comment]
Let's just go ahead and fix it. It's still a bit unclear where the refactor will begin. Thank you so much for confirming!

**Comment by eh2077** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

An appropriate message is not shown when inviting a member that is already invited

### What is the root cause of that problem?

The root cause of this issue is that when we format a header message, we don't check if the search term is an invited login(phone or email) or not, see

https://github.com/Expensify/App/blob/820fa93ad1e799c12d8b639b98018b5350ea17af/src/pages/workspace/WorkspaceInvitePage.js#L261

### What changes do you think we should make in order to solve the problem?

To fix this issue, we can 

1. Define and locale key, like `userIsInvitedToWorkspaceMessage`,  in `en.js` and `es.js` for the message to hint that the search term is invited, like

   ```
   userIsInvitedToWorkspaceMessage: ({login, workspace}) => `${login} is already a member of ${workspace}`
   ```

   We can discuss with team to determine the hint message and Spanish translation.

2. Check if the search term is an invited login or not. We can modify this line to

https://github.com/Expensify/App/blob/820fa93ad1e799c12d8b639b98018b5350ea17af/src/pages/workspace/WorkspaceInvitePage.js#L261

to

```js
const excludeLogins = _.difference(this.getExcludedUsers(), [...CONST.EXPENSIFY_EMAILS]);
const headerMessage = _.contains(excludeLogins, this.state.searchTerm) 
    ? Localize.translate(preferredLocale, 'common.userInvited', {login: this.state.userToInvite.login, workspace: policyName})
    : OptionsListUtils.getHeaderMessage(this.state.personalDetails.length !== 0, Boolean(this.state.userToInvite), this.state.searchTerm);
```

### What alternative solutions did you explore? (Optional)

We can also extend method `getHeaderMessage`

https://github.com/Expensify/App/blob/820fa93ad1e799c12d8b639b98018b5350ea17af/src/libs/OptionsListUtils.js#L986

by adding parameter `invitedUsers` 

```js
/**
 * Helper method that returns the text to be used for the header's message and title (if any)
 *
 * @param {Boolean} hasSelectableOptions
 * @param {Boolean} hasUserToInvite
 * @param {String} searchValue
 * @param {Array} invitedUsers
 * @param {Boolean} [maxParticipantsReached]
 * @return {String}
 */
function getHeaderMessage(hasSelectableOptions, hasUserToInvite, searchValue, invitedUsers = [], maxParticipantsReached = false) {
    if (maxParticipantsReached) {
        return Localize.translate(preferredLocale, 'common.maxParticipantsReached', {count: CONST.REPORT.MAXIMUM_PARTICIPANTS});
    }
    
    // check if invitedUsers contains searchValue
    if (_.contains(invitedUsers, searchValue)) {
        return Localize.translate(preferredLocale, 'common.userInvited');
    }
  

```

**Comment by eh2077** [comment]
### Proposal
[Updated](https://github.com/Expensify/App/issues/21193#issuecomment-1617945832) to solve two instances together.

1. If searching an already invited login(email/phone), then we display 
    ```
    ${login} is already a member of ${workspace}
    ```
2. If searching an excluded Expensify email, like `concierge@expensify.com`, then we display 
    ```
    It's not allowed to invite Expensify email ${email} to ${workspace}
    ```

@mallenexpensify Would you mind sharing your thoughts on the above two hint messages to be added?

cc @fedirjh

**Comment by mallenexpensify** [comment]
@fedirjh , I'm a lil confused/unsure here. 
> Actually the excluded emails are already invited one (which is reported in this issue) + expensify emails ([concierge@expensify.com](https://www.expensify.com/_support/?supportEmail=concierge%40expensify.com&comment=better%20supportal%20issue%20nubmer%3A%2021193&location=_support/), [help@expensify.com](https://www.expensify.com/_support/?supportEmail=help%40expensify.com&comment=better%20supportal%20issue%20nubmer%3A%2021193&location=_support/) ...) , I think we should have another result for that.

and

> Already invited members is an instance, I think we should fix all instances , and it looks that we have two instances already invited and expensify emails.

Can you rephrase your question?  Which emails or domains do we specifically need a different message for?

**Comment by mallenexpensify** [comment]
> should display another message if we try to add one of the excluded Expensify emails, or should we simply display "not found" ?

Good question @fedirjh ,   Let's go with "Invalid email address" since that's technically what it is and I can't think of better wording

**Comment by pecanoro** [comment]
Ok, I went through all proposals.

@tienifr Your solution seems to complicate things unnecessarily, dealing with an associative array for the reason is a bit overkill.

@spcheema @eh2077 It's true these two solutions are really similar, but I also don't see the point in maintaining another list just for the invited users. @spcheema Why do you think it makes complexity worse? Unless there are any performance issues I am missing, I completely agree with @fedirjh reasoning here.

**Comment by pecanoro** [comment]
@spcheema Unless I am missing something, if I remember correctly from my algorithms class, the time complexity between `O(n)` and `2 * O(n)` is the same. There are more details in this link about the reasoning behind it: https://www.reddit.com/r/algorithms/comments/wrdm3v/is_there_any_significant_difference_between_on/

**Comment by pecanoro** [comment]
Honestly, after all these discussions, @fedirjh alternative solution is the best one. We don't need a `_.difference`, an extra `alreadyInvitedUser` or a new structure. The `if else` should be enough:

```
if (CONST.EXPENSIFY_EMAILS.includes(this.state.searchTerm.trim())) {
    headerMessage = '...'
} else if (excludedUsers.includes(this.state.searchTerm.trim())) {
    headerMessage = '...'
} else {
    headerMessage = OptionsListUtils.getHeaderMessage...;
}
```

Since it's an if-else, getting on the first condition means it won't try to evaluate the second one, so that should be everything that we need. The question is "who will fix it?" 😄

**Comment by eh2077** [comment]
@pecanoro Thanks for reviewing proposals! As @fedirjh mentioned https://github.com/Expensify/App/issues/21193#issuecomment-1638523643 he made the suggestion based on my proposal, so I'm happy to make the PR to move this issue forward!

**Comment by pecanoro** [comment]
I will get back to you all later today after we made a decision since it's a tricky case.

**Comment by pecanoro** [comment]
We discussed it and decided that the fair thing to do is to go with @eh2077 since @fedirjh suggestion was based on this proposal. However, since we recognize that all approaches were similar and inspired by each other's, we will do 25% for @spcheema and 75% for @eh2077 with @eh2077 being the one raising the PR. Does it sound good?

**Comment by pecanoro** [comment]
@spcheema The decision is already made so let's just keep moving with the issue forward.

**Comment by mallenexpensify** [comment]
Sorry for the delay.. 
1. The original job closed, @daveSeife is hired for reporting and so is @eh2077 (who will be compensated 75%) . @fedirjh , can you accept the offer please? https://www.upwork.com/jobs/~0129bb1fb0fcd4cb51
2. @spcheema , can you apply to this new job please and reply here once you have? https://www.upwork.com/jobs/~01f5b4bf8537c32990  Pay will be $250 for you

**Comment by mallenexpensify** [comment]
@spcheema , can you please accept the job and reply here once you have? 
https://www.upwork.com/jobs/~01f5b4bf8537c32990

Note to self... 
Issue reporter: @daveSeife paid $250 via Upwork (paid early by mistake

**Comment by pecanoro** [comment]
@mallenexpensify What's the status here when it comes to the payments? @fedirjh Can you please complete the checklist?

**Comment by mallenexpensify** [comment]
Issue reporter: @daveSeife paid $250 via Upwork  
Contributor: @spcheema paid $250 via Upwork
Contributor: @eh2077 paid $750 via Upwork
Contributor+: @fedirjh paid $1000 via Upwork 

This issue is an anomaly @spcheema was paid 25% and @eh2077 was paid 75%, more details [here](https://github.com/Expensify/App/issues/21193#issuecomment-1647772677) 

TR GH: https://github.com/Expensify/Expensify/issues/309331
